### PR TITLE
Enhancement of one-shot NAS (v2.9)

### PR DESCRIPTION
**Pull request by ultmaster**
#### Features

* Add export_probs to monitor the architecture weights.
* Rewrite configure_optimizers, functions to step optimzers / schedulers, along with other hooks for simplicity, and to be compatible with latest lightning (v1.7).
* Support gradient clipping in DARTS.
* Support arbitary arguments in random one-shot, because it's automatic optimization.
* Support +1 (none) in differentiable cell, to mimic the input choice (similar to what DARTS did).
* Reimplement ProxylessNAS with forward/backward hooks. This increases its robustness and applicability.

#### Bug fixes

* Fix a misuse of lambda expression in NAS-Bench-201 (used in tests).
* Fix the gumbel temperature (the original one doesn't work at all).
* Fix an issue that formats tensors in `weighted_sum`, making it very slow.
* Fix the label sharing in differentiable strategies. The original one wasn't sharing at all.
* Fix the memo reuse in exporting differentiable cell.

#### Misc

* Add a more friendly error message when no architecture weights are available.
* Added some more explanations in one-shot documentation: manual optimization, some extra details in one-shot supermodules, typo fixes.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### How to test ###

In another PR.

### CODE REVIEW

File: nni/nas/oneshot/pytorch/supermodule/differentiable.py [hunk 1]
```diff
@@ -376,6 +404,16 @@ def export(self, memo):
             result[name] = spec.values[chosen_index]
         return result
 
+    def export_probs(self, memo):
```
Comment (J-shang) [inline nni/nas/oneshot/pytorch/supermodule/differentiable.py:407]: seems the `export_probs` logic are similar? and some have type hints and some not, anyway, it's not matter.

Comment (ultmaster) [reply nni/nas/oneshot/pytorch/supermodule/differentiable.py:407]: Yes. You might notice all DifferentiableXXX are very similar, not only `export_probs`. The framework design is to blame.

File: nni/nas/oneshot/pytorch/base_lightning.py [hunk 2]
```diff
@@ -295,236 +309,229 @@ def export(self) -> dict[str, Any]:
...
+        if isinstance(optim_conf, dict):
+            return [optim_conf] + [{'optimizer': optimizer} for optimizer in arch_optimizers]
+        # 4. multiple dictionaries
+        if isinstance(optim_conf, (list, tuple)) and all(isinstance(d, dict) for d in optim_conf):
+            return list(optim_conf) + [{'optimizer': optimizer} for optimizer in arch_optimizers]
+        # 5. single list or tuple, multiple optimizer
+        if isinstance(optim_conf, (list, tuple)) and all(isinstance(opt, Optimizer) for opt in optim_conf):
+            return list(optim_conf) + arch_optimizers
+        # unknown configuration
+        warnings.warn('Unknown optimizer configuration. Architecture optimizers will be ignored. Strategy might fail.', UserWarning)
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply nni/nas/oneshot/pytorch/base_lightning.py:380]: Good point.

File: nni/nas/oneshot/pytorch/base_lightning.py [hunk 3]
```diff
@@ -295,236 +309,229 @@ def export(self) -> dict[str, Any]:
...
-        ----------
-        arc_optimizers : list[Optimizer], Optimizer
-            Optimizers used by a specific NAS algorithm. Return None if no architecture optimizers are needed.
+        -------
+        Optimizers used by a specific NAS algorithm. Return None if no architecture optimizers are needed.
         """
         return None
 
-    def call_lr_schedulers(self, batch_index):
+    def advance_optimizers(
```
[the remark this answers is not in the document]
Comment (ultmaster) [reply nni/nas/oneshot/pytorch/base_lightning.py:410]: Make sense.

File: nni/nas/oneshot/pytorch/supermodule/proxyless.py [hunk 4]
```diff
@@ -21,75 +21,148 @@
...
+        for t in tensor:
+            yield from _iter_tensors(t)
+    elif isinstance(tensor, dict):
+        for t in tensor.values():
+            yield from _iter_tensors(t)
+
+
+def _pack_as_tuple(tensor: Any) -> Any:
+    """Return a tuple of tensor with only one element if tensor it's not a tuple."""
+    if not isinstance(tensor, tuple):
```
Comment (J-shang) [inline nni/nas/oneshot/pytorch/supermodule/proxyless.py:54]: should this place consider `list`?

**Review by J-shang — APPROVED** [review]

**Review by QuanluZhang — APPROVED** [review]
